### PR TITLE
Link ExtraCoords to the parent Cube

### LINKED ISSUE #349: Extra Coords not removed by NDCube slicing when their axis is sliced away
### Description

NDCube slicing does not remove extra coords if their axis is sliced away.  Instead it is moved to the axis in the sliced cube with the equivalent axis number.

### Steps to Reproduce

```python
>>> import numpy as np
>>> import astropy.wcs
>>> data = np.ones((3, 4, 5))
>>> wcs_input_dict = {
...     'CTYPE1': 'WAVE    ', 'CUNIT1': 'Angstrom', 'CDELT1': 0.2, 'CRPIX1': 0, 'CRVAL1': 10, 'NAXIS1': 5,
...     'CTYPE2': 'HPLT-TAN', 'CUNIT2': 'deg', 'CDELT2': 0.5, 'CRPIX2': 2, 'CRVAL2': 0.5, 'NAXIS2': 4,
...     'CTYPE3': 'HPLN-TAN', 'CUNIT3': 'deg', 'CDELT3': 0.4, 'CRPIX3': 2, 'CRVAL3': 1, 'NAXIS3': 3}
>>> input_wcs = astropy.wcs.WCS(wcs_input_dict)

>>> from astropy.time import Time, TimeDelta
>>> base_time = Time('2000-01-01', format='fits', scale='utc')
>>> timestamps = Time([base_time + TimeDelta(60 * i, format='sec') for i in range(data.shape[0])])

>>> from ndcube import ExtraCoords
>>> my_extra_coords = ExtraCoords()
>>> my_extra_coords.add_coordinate('time', (0,), timestamps)

>>> from ndcube import NDCube
>>> my_cube = NDCube(data, input_wcs, extra_coords=my_extra_coords)

>>> my_cube.array_axis_physical_types
[('custom:pos.helioprojective.lat', 'custom:pos.helioprojective.lon'),
 ('custom:pos.helioprojective.lat', 'custom:pos.helioprojective.lon'),
 ('em.wl', 'time')]

>>> my_cube[:, :, 0].array_axis_physical_types
[('custom:pos.helioprojective.lat', 'custom:pos.helioprojective.lon'),
 ('custom:pos.helioprojective.lat', 'custom:pos.helioprojective.lon', 'time')]
```

### System Details

- SunPy Version:
- Astropy Version:
- Python Version:
- OS information:

### LINKED ISSUE #408: Supply global_coords during NDCube Initialization?
### Description

In order to stay consistent with `extra_coords`, should it be possible to supply a `GlobalCoords` instance while initializing and `NDCube`?  If so, what should the API look like?

Currently when an `NDCube` is initialized, an empty `GlobalCoords` instance is created and attached via the `extra_coords` attribute.  Global coords must then be added subsequently.

### LINKED ISSUE #417: Fix the ordering of array_coordinates when adding ExtraCoords
Ensure that array_dimensions is in array order not world order in `ExtraCoords.add`

### PR DESCRIPTION
**Pull request by Cadair**
Once upon a time I decided to throw up a quick fix for #349. Then I realised that to fix that issue the `ExtraCoords` object needed to  know the dimensionality of the array, which means it needs a reference to the cube like `GlobalCoords`. So this PR addresses this by making the following changes:

* Change the constructor of `ExtraCoords` to take one optional argument `ndcube`.
* Remove the extra_coords kwarg from `NDCube` the canonical way to add coords to the `ExtraCoords` or `GlobalCoords` is `NDCube.xxxx_coords.add`.
* Write a descriptor for the `_extra_coords` and `_global_coords` attributes which when a coords object is assigned to the attribute sets the `._ndcube` attribute of the object. This means that all our internal code which sets the hidden attributes automatically ensures the parent reference is correct.
* Fix the cause of #349 which was `ExtraCoords.mapping` using array coordinates not pixel coordinates. Add a regression test for this as well.

This therefore fixes #349 and also fixes #417

It also closes #408 by deciding no for both `GlobalCoords` and `ExtraCoords`; however, it makes it real easy to change our minds in the future, by changing the descriptors to be not prefixed with `_`.

### CODE REVIEW

**Comment by Cadair** [comment]
Some of the tests are janky because of gWCS versions and fixes... I will investigate.

**Comment by Cadair** [comment]
@DanRyanIrish This is good to go when you get a chance.

File: ndcube/ndcube.py [hunk 1]
```diff
@@ -128,6 +128,38 @@ def crop_by_values(self, lower_corner, upper_corner, units=None, wcs=None):
...
+
+        return getattr(obj, self._attribute_name)
+
+    def __set__(self, obj, value):
+        if isinstance(value, self._default_type):
+            value._ndcube = obj
+        elif issubclass(value, self._default_type):
+            value = value(obj)
+        else:
+            raise ValueError("wat")
```
Comment (DanRyanIrish) [inline ndcube/ndcube.py:158]: More helpful error meesage here ;)

File: ndcube/ndcube.py [hunk 2]
```diff
@@ -128,6 +128,38 @@ def crop_by_values(self, lower_corner, upper_corner, units=None, wcs=None):
...
+    """
+    A descriptor which gives the property a reference to the cube to which it is attached.
+    """
+    def __init__(self, default_type):
+        self._default_type = default_type
+        self._property_name = None
+
+    def __set_name__(self, owner, name):
+        self._property_name = name
+        self._attribute_name = f"_{name}"
```
Comment (DanRyanIrish) [inline ndcube/ndcube.py:141]: Might be helpful to add a comment here explaining different between `_property_name` and `_attribute_name`.

File: ndcube/ndcube.py [hunk 3]
```diff
@@ -174,9 +206,11 @@ class NDCubeBase(NDCubeSlicingMixin, NDCubeABC):
         Default is False.
 
     """
+    _extra_coords = NDCubeLinkedDescriptor(ExtraCoords)
```
Comment (DanRyanIrish) [inline ndcube/ndcube.py:209]: Add comment here like, `# Initializing empty ExtraCoords and GlobalCoords instances on NDCube.`

By the by, why are these lines floating on their own and not in `__init__`.  Is it for the same reason as in astropy modeling (whatever that reason actually is?)

**Review by DanRyanIrish — APPROVED** [review]
Approved pending a couple minor comments

**Comment by DanRyanIrish** [comment]
Does this fix #417 as well?

File: ndcube/ndcube.py (hunk 3 above)
Comment (Cadair) [reply ndcube/ndcube.py:209]: These are class level attributes not instance level attributes (as they would be if they were `self.whatever` in the constructor). All instances of this class share the one instance of this descriptor object.

**Pull request merged**